###### R script to visualize the most abundant microsatellite alleles from Illumina amplicon sequencing and the generation of a matrix based on allele
###### length and frequency distribution, generates a color-coded pdf with barplots of allele lengths and their relative frequency  
###### generated by S. Winter

## check commented lines number 54 for alpha now set to 0.62 and lines 84 and 103 for x-axis limits

#install.packages("reshape")
#install.packages("reshape2")
#library(reshape2)
if(!require(reshape2)){install.packages("reshape") & install.packages("reshape2")}   #needed for command line call in bash script

#Passing over parameters from command line/ bash script
#cat(getwd(), "\n")
#setwd("/home/swinter/Igel_Microsat")
args <- commandArgs(TRUE)
path=(as.character(args[1]))        #path of directory of inputfiles with allele lengths and counts per marker and sample
mincount=as.numeric(args[2])        #minimum treshold of total read count to consider that marker in sample, absolute minimum 500 recommended, better higher
pathPDF=(as.character(args[3]))     #path to directory where to store pdf 
alpha=as.numeric(args[4])           #threshold for relative frequency per read length to be considered as homozygous
pathmatrix=(as.character(args[5]))     #path to directory where to store marker matrix 
pathcsvlist=(as.character(args[6]))   #path to directory where to store list of markers and samples with read length
a=(as.numeric(args[7]))             # lower x-axis limit for plots
b=(as.numeric(args[8]))          #upper x-axis limit for plots

#define if the right input parameters have been provided - only necessary for mincount and alpha
#mincount=400
#if (is.numeric(mincount) != TRUE | (mincount>20) != TRUE) {stop("You have not provided a minimum number for read counts per marker and sample or
                                                             #       the number is too small. It should be larger than 20.")}
#alpha=0.4
if (is.numeric(alpha) != TRUE | (alpha<0.9) != TRUE | (alpha>0.5) != TRUE) {
  stop("You have not provided an adequate number for alpha, should not be smaller than 0.5 and not larger than 0.9.")}


#possibly enable facultative command line parameters for xlim upper and lower for plots
#a=as.numeric(args[5]) 
#b=as.numeric(args[6])
#a=100; b=950
#if (is.numeric(a) != TRUE | (a>500) != TRUE | (a<100) != TRUE) {
#  stop("You have not provided an adequate number for the lower x-axis limit above 100 bp read length and below 500 bp read length")}
#if (is.numeric(b) != TRUE | (b>900) != TRUE | (b<200) != TRUE) {
#  stop("You have not provided an adequate number for the upper x-axis limit above 200 bp read length and below 900 bp read length")}

#hard coded paths for developing the script:
##cat("path: ", path, "\n")
#path="~/Igel_Microsat/Outputdir17-09-12/MarkerStatistics/"
#pathPDF = "~/Igel_Microsat/Outputdir17-09-12/Markerplots/markerplots.pdf"
#pathmatrix = "~/Igel_Microsat/Outputdir17-09-12/Markerplots/markermatrix.csv"

#Generate pdf with bar charts per marker and sample, generate 6 plots per page 
pdf(file=pathPDF,width=7,height=18)
par(mfrow=c(6,1))
op <- options(stringsAsFactors=F) 
dat<-data.frame()    #initialise empty dataframe for marker length information

#file = dir(path,full.names=TRUE)[3]
#alpha=0.62             #define alpha treshold of relative frequency for homozygous allele - should be commented
i=0
#a=300
#b=500
# for (file in dir(path,full.names=TRUE)) {
#   #print(paste ("filename is ", file))
#   if (file.size(file) == 0 ) {                           # if (file.info(file)$size == 0 )
#     print(paste("empty files are: ", i, file,sep=" "))
#    # filein<-read.table(file,header=F,sep= " ")
#     i=i+1
#   }
# }
#file="/home/swinter/Igel_Microsat/Outputdir17-09-07/MarkerStatistics//HH10_AAAG_IBK9_.statistics"

for (file in dir(path,full.names=TRUE)) {
  print(paste ("filename is ", file))
  output<-basename(file.path(file,fsep=".statistics")); output
  print(paste ("basename is", output))
  stringsplit1<-sapply(strsplit(output, split = "_.statistics"),"[",1); stringsplit1    #extract part of file name with sample name and marker name for matrix
  splits<-as.character(stringsplit1); splits
  repstring<-sapply(strsplit(splits, split = "_"), "[", 2); repstring
  samplename<-sapply(strsplit(splits, split = "_"), "[", 3); samplename
  markern<-sapply(strsplit(splits, split = "_"), "[", 1); markern   #Regexpression to extract markername
  #markern1<-substring(markern, 2); markern1
  markername<-paste(markern,repstring,sep="_"); markername
  repsiz1<-nchar(repstring); repsiz1                    #extract repeat size length from repeat motif
  repsiz<-as.integer(repsiz1); repsiz
  if (file.size(file) == 0 ) {   
    print(paste("empty files are: ", i, file,sep=" "))
    i=i+1
    plot(1,ylab="empty or nearly empty file", xlab="", xlim=c(a, b), ylim=c(0,105), type="h", main=output, sub=paste("Total read count is",sep=" "))
    mtext("empty file",side=3,cex=0.8)
    foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = "empty")
    dat<-rbind(dat, foo)
    foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = "empty")
    dat<-rbind(dat, foo)
  } else {   
  filein<-read.table(file,header=F,sep= " "); filein
  maxcount<-sum(filein[,2]) #; maxcount   #maxcount = total read count
  maxPeak<-max(filein[,2]); maxPeak 
  #maxleng<-filein$V1[filein$V2==maxPeak]; maxleng
  maxleng<-tail(filein$V1,n=1); maxleng
  #maxPeak2<-sort(filein$V2,decreasing=TRUE)[n-1]; maxPeak2
  maxPeak2<-head(tail(filein$V2,n=2),n=1); maxPeak2
  #maxleng2<-filein$V1[filein$V2==maxPeak2]; maxleng2             #Problem zwei ganz gleich häufige allele
  maxleng2<-head(tail(filein$V1,n=2),n=1); maxleng2
  n<-length(filein[,2]); n
  #rbPal<-colorRampPalette(c('red','blue'))          #load colour palette
  #filein$Col<-rbPal(100)[as.numeric(cut(filein$V2,breaks = 100))]    #create colour palette with red colour at higher peaks
  #plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(400,550),ylim=c(0,100), col= filein$Col) 
  #plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,100), col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"red", "black"))
  #  plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=maxcount,lwd = 5, xlim=c(a,b),ylim=c(0,100), col= filein$Col) 
  allelenames<-as.character(filein[,1]); allelenames
  countnames<-as.character(filein[,2]); countnames
  #text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames)
  if ((maxcount > 20) & (n > 1)) {                               #call mincount variable from command line input here
          #first case one length highly abundant - homozygous
              if ( maxPeak/maxcount >= alpha) {
                  plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,105), col=ifelse(filein$V1==maxleng, "blue", "black"))
                  text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames)
                  mtext(paste("homozygous read count and allele is:", maxleng, ":", maxPeak, sep=" "),side=3,cex=0.8, col="blue")
                  foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                  dat<-rbind(dat, foo)
                  foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng)
                  dat<-rbind(dat, foo)
          #heterozygous variants
                  #second case: heterozygous, most abundant and second most abundant bigger than alpha, difference in length according to repeat length, second highest allele longer than highest
              } else if ( ((maxPeak/maxcount + maxPeak2/maxcount) >= alpha) & ((abs(maxleng2-maxleng)) >= repsiz) & ((maxleng2-maxleng) > 0) & ((abs(maxleng2-maxleng)) %% repsiz == 0)) {    #zweitgrößtes Allel länger als erstes 
                      plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,105), col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"green", "black"))
                      mtext(paste("heterozygous, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), side=3,col="dark green",cex=0.8) 
                      text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames, col="dark green")
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                      dat<-rbind(dat, foo)
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
                      dat<-rbind(dat, foo)
      
                  #third case - heterozygous, diff most abundant and second most in length more than 1 x repeat size (exclusion stutter), difference exactly integer x repeat size
              } else if ( ((maxPeak/maxcount + maxPeak2/maxcount) >= alpha) & (abs(maxleng2-maxleng) > repsiz) & (abs(maxleng2-maxleng) %% repsiz == 0) ) {
                      plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,105), col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"green", "black"))
                      mtext(paste("heterozygous, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), cex=0.8, side=3, col="dark green") 
                      text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames, col="dark green")
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                      dat<-rbind(dat, foo)
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
                      dat<-rbind(dat, foo)
          
                  # fourth case - point mutations that are minimum size of 60% max peak but don't match repeat size  
              } else if ( ((maxPeak/maxcount + maxPeak2/maxcount) >= alpha) & ((maxPeak2/maxcount) > (0.6*maxPeak/maxcount)) & ((abs(maxleng2-maxleng)) %% repsiz != 0) )   {  #second peak needs to minimum read count of 0.6xhighest peak                                                
                      plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,105), col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"green", "black"))
                      mtext(paste("possibly heterozygous point mutation, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), cex=0.8, side=3,col="green")
                      text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames, col="green")
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                      dat<-rbind(dat, foo)
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
                      dat<-rbind(dat, foo)
                 # fifth case - only one repeat size difference but second largest peak is minimum size of 0.75 % of the largest one,    
              } else if ( ((maxPeak/maxcount + maxPeak2/maxcount) >= alpha) & ((maxPeak2/maxcount) > (0.75*maxPeak/maxcount)) & (abs(maxleng2-maxleng) == repsiz) )   { 
                      plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,105), col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"green", "black"))
                      mtext(paste("heterozygous, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), cex=0.8, side=3, col="dark green") 
                      text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames, col="dark green")
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
                      dat<-rbind(dat, foo)
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
                      dat<-rbind(dat, foo)
                      
              } else   {
                      plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,105), col=ifelse(filein$V1==maxleng | filein$V1==maxleng2,"red", "black"))
                      mtext(paste("needs manual check, read counts of highest peaks are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "),side=3,col="grey", cex=0.8)
                      text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames, col="red")
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = paste(maxleng, "man check", sep="_"))
                      dat<-rbind(dat, foo)
                      foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = paste(maxleng2, "man check", sep="_"))
                      dat<-rbind(dat, foo)
              }        
  } else if ((maxcount > 20) & (n = 1)) { 
        plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,105), col=ifelse(filein$V1==maxleng,"blue", "black"))
        mtext(paste("Homozygous, only one read length present:", maxleng, ":", maxPeak, sep=" "),side=3,col="blue", cex=0.8)
        text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames, col="blue")
        foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
        dat<-rbind(dat, foo)
        foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng)
        dat<-rbind(dat, foo)
  } else if (maxcount == 0) {
        plot(1,ylab="empty file", xlab="", xlim=c(a, b), ylim=c(0,105), type="h", main=output, sub=paste("Total read count is",sep=" "))
        mtext("empty file",side=3,cex=0.8)
        foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = "empty")
        dat<-rbind(dat, foo)
        foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = "empty")
        dat<-rbind(dat, foo)
  } else {
     plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,105), col="grey")
     mtext(paste("too little read count:", maxleng, ":", maxPeak, sep=" "),side=3,cex=0.8, col="grey")
     text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames, col="grey")
     foo<-data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = "too little reads")
     dat<-rbind(dat, foo)
     foo<-data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = "too little reads")
     dat<-rbind(dat, foo)
  }
  }
}
dev.off()

dat
dat<- melt(dat,id.vars= c("samplename","marker"))
res<-dcast(dat,samplename~marker,function(x){x[1]})
res
write.csv(res,file=pathmatrix,row.names=FALSE)

options(op)  #wegen Fehlermeldungen nötig: 0: In `[<-.factor`(`*tmp*`, ri, value = 437L) : ungültiges Faktorniveau, NA erzeugt

#KOPIE alter for Schleife mit IF ELSE
  
#   for (file in dir(path,full.names=TRUE)) {
#     print(paste ("filename is ", file))
#     filein<-read.table(file,header=F,sep= " ")
#     output<-basename(file.path(file,fsep=".statistics"))
#     maxcount<-sum(filein[,2])                           #maxcount = total read count
#     if ((maxcount > 99) & (maxPeak > 30)) {                               #call mincount variable from command line input here
#       maxPeak<-max(filein[,2])                          #find maximal read count
#       n<-length(filein[,2])
#       maxPeak2<-sort(filein[,2],partial=n-1)[n-1]       #find second highest read count                                    
#       print(maxPeak);print(maxPeak2)
#       rbPal<-colorRampPalette(c('red','blue'))          #load colour palette
#       filein$Col<-rbPal(100)[as.numeric(cut(filein$V2,breaks = 100))]    #create colour palette with red colour at higher peaks
#       plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=maxcount,lwd = 5, xlim=c(400,500),ylim=c(0,100), col= filein$Col) 
#       #textmaxpeak<-parse(text=paste(maxPeak, "*~Read count of max Peak"))
#       #mtext(maxPeak,side=4) #check ob das geht
#       #if input defined regarding x axis put in variables a and b
#       #  a=400; b=500
#       #   plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", main=output, sub=maxcount,lwd = 5, xlim=c(a,b),ylim=c(0,100), col= filein$Col) 
#       allelenames<-as.character(filein[,1]); allelenames
#       countnames<-as.character(filein[,2]); countnames
#       text(filein[,1]-2, 100*filein[,2]/maxcount+5 ,allelenames)
#       maxleng<-filein$V1[filein$V2==maxPeak]
#       maxleng2<-filein$V1[filein$V2==maxPeak2]
#       splits<-strsplit(file, split = "Statistics/")
#       stringsplit1<-sapply(strsplit(file, split = "Statistics/"),"[",2)    #extract part of file name with sample name and marker name for matrix
#       splits<-as.character(stringsplit1)
#       splits2<-strsplit(splits, split = "_")
#       repstring<-sapply(strsplit(splits, split = "_"), "[", 2)
#       samplename<-sapply(strsplit(splits, split = "_"), "[", 3)
#       markern<-sapply(strsplit(splits, split = "_"), "[", 1)   #Regexpression to extract markername
#       substring(markern, 1, 1)
#       markern1<-substring(markern, 2) 
#       markername<-paste(markern1,repstring,sep="_")
#       print(paste("Markername ist", markername))
#       repsiz<-nchar(repstring)                    #extract repeat size length from repeat motif
#       repsiz<-as.integer(repsiz)
#   if ( maxPeak/maxcount >= alpha) {
#   mtext(paste("homozygous read count and allele is:", maxleng, ":", maxPeak, sep=" "),side=3)
#   foo <- data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
#   dat <- rbind(dat, foo)
#   foo <- data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng)
#   dat <- rbind(dat, foo)
#   
#   #second case - heterozygous - most abundant and second most abundant bigger than alpha, difference in length according to repeat length, second highest allele longer than highest  
# } else if ( (maxPeak/maxcount + maxPeak2/maxcount >= alpha) & ((abs(maxleng2-maxleng)) >= repsiz) & ((maxleng2-maxleng) > 0) ) {    #zweitgrößtes Allel länger als erstes 
#   mtext(paste("heterozygous, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), side=3,col="dark green") 
#   foo <- data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
#   dat <- rbind(dat, foo)
#   foo <- data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
#   dat <- rbind(dat, foo)
#   
#   #third case - heterozygous -   most abundant and second most abundant bigger than alpha, diff most abundant and second most in length more than 1 x repeat size (exclusion stutter), difference exactly integer x repeat size
# } else if ( (maxPeak/maxcount + maxPeak2/maxcount >= alpha) & ((abs(maxleng2-maxleng)) > repsiz) & ((abs(maxleng2-maxleng)) %% repsiz == 0) ) {
#   mtext(paste("heterozygous, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), side=3, col="dark green") 
#   foo <- data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
#   dat <- rbind(dat, foo)
#   foo <- data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
#   dat <- rbind(dat, foo)
#   
#   #point mutations that are minimum size of 60% max peak but don't match repeat size  
# } else if ( (maxPeak/maxcount + maxPeak2/maxcount >= alpha) & (maxPeak2/maxcount > (0.6*maxPeak/maxcount)) & ((abs(maxleng2-maxleng)) %% repsiz != 0) )   {  #second peak needs to minimum read count of 0.6xhighest peak                                                
#   mtext(paste("possibly heterozygous point mutation, size and read counts of the alleles are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "), side=3,col="green")
#   foo <- data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = maxleng)
#   dat <- rbind(dat, foo)
#   foo <- data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = maxleng2)
#   dat <- rbind(dat, foo)
#   
# } else
#   mtext(paste("needs manual check, read counts of highest peaks are:", maxleng, ":", maxPeak, ";", maxleng2, ":", maxPeak2, sep=" "),side=3,col="red")
# foo <- data.frame(samplename = samplename, marker = paste(markername, "1", sep="_"), maxleng = paste(maxleng, "man check", sep="_"))
# dat <- rbind(dat, foo)
# foo <- data.frame(samplename = samplename, marker = paste(markername, "2", sep="_"), maxleng = paste(maxleng2, "man check", sep="_"))
# dat <- rbind(dat, foo)
# }
# }
# dev.off()

# a=400
# b=550
# file="/home/swinter/Igel_Microsat/Outputdir17-09-07/MarkerStatistics//HH23_CCT_KLF74_.statistics"
# filein<-read.table(file,header=F,sep= " "); filein
# maxcount<-sum(filein[,2]); maxcount   #maxcount = total read count
# maxPeak<-max(filein[,2]); maxPeak 
# maxleng<-tail(filein$V1,n=1); maxleng  #WOULD LIKE to colour that in figure
# maxPeak2<-head(tail(filein$V2,n=2),n=1); maxPeak2             #Problem zwei ganz gleich häufige allele
# maxleng2<-head(tail(filein$V1,n=2),n=1); maxleng2
# rbPal<-colorRampPalette(c('red','blue'))          #load colour palette
# filein$Col<-rbPal(100)[as.numeric(cut(filein$V2,breaks = 100))] 
# plot(filein[,1],100*filein[,2]/maxcount,xlab="fragment length",ylab="absolute read count",type="h", sub=paste("Total read count is",maxcount,sep=" "),lwd = 5, xlim=c(a,b),ylim=c(0,100), col=ifelse(filein$V1==maxleng | filein$V2==maxleng2,"red", "black")) # filein$Col) 
